lazy_dataset/core.py

```diff
@@ -1861,6 +1885,82 @@ def catcher(index):
                 else:
                     yield data
 
+    def _single_thread_prefetch(self):
+        """
+                >>> import string
+        >>> ascii = string.ascii_lowercase
+        >>> ds = DictDataset({k: v for v, k in enumerate(ascii[:5])})
+        >>> list(ds)
+        [0, 1, 2, 3, 4]
+        >>> def foo(ex):
+        ...     print(f'called with {ex}')
+        ...     return ex
+        >>> ds = ds.map(foo)
+        >>> ds = PrefetchDataset(ds.filter(lambda x: x != 2), 1, 2)
+        >>> next(iter(ds))  # The buffer is filled with [1, 3]. 2 is ignored.
+        called with 0
+        called with 1
+        called with 2
+        called with 3
+        0
+        >>> list(ds)  # Note: 2 is missing
+        called with 0
+        called with 1
+        called with 2
+        called with 3
+        called with 4
+        [0, 1, 3, 4]
+        """
+        import threading
+        import queue
+
+        if self.catch_filter_exception:
+            if self.catch_filter_exception is True:
+                exceptions = FilterException
+            else:
+                exceptions = self.catch_filter_exception
+            input_dataset = CatchExceptionDataset(
+                self.input_dataset,
+                exceptions=exceptions,
+            )
+        else:
+            input_dataset = self.input_dataset
+
+        shutdown = False
+        data_queue = queue.Queue(self.buffer_size)
+
+        def worker():
+            if shutdown:
+                return
+            try:
+                for ex in input_dataset:
+                    if shutdown:
+                        return
+                    data_queue.put(ex)
+                    if shutdown:
+                        return
+            finally:
+                data_queue.put(unique_object)
+
+        unique_object = object()
```

Comment:
Can you move this definition before `worker`? It is used in `worker`

Reply:
Done